src/commands/up.rs

```diff
@@ -189,10 +193,25 @@ impl UpCommand {
         if self.build {
             app_source.build().await?;
         }
-
         let mut locked_app = self
             .load_resolved_app_source(resolved_app_source, &working_dir)
-            .await?;
+            .await
+            .context("Failed to load application")?;
+        if !self.components.is_empty() {
+            retain_components(&mut locked_app, &self.components)?;
```

Comment:
I would add some additional context to this function and some of the subfunctions with `.context()`. Right now an error from this function might simply read "failed to get allowed hosts" which is highly confusing without the context of the code.